src/hockeypuck/hkp/handler.go

```diff
@@ -216,18 +216,23 @@ func (h *Handler) HashQuery(w http.ResponseWriter, r *http.Request, _ httprouter
 		return
 	}
 	var result []*openpgp.PrimaryKey
+	digests := make(map[string][]*openpgp.PrimaryKey)
+
 	responseLen := 0
 	for _, digest := range hq.Digests {
-		rfps, err := h.storage.MatchMD5([]string{digest})
-		if err != nil {
-			log.Errorf("error resolving hashquery digest %q", digest)
-			continue
-		}
-		keys, err := h.storage.FetchKeys(rfps)
-		if err != nil {
-			log.Errorf("error fetching hashquery key %q", digest)
-			continue
+		// Check the already processed digests for duplicates
+		keys := digests[digest]
+
+		if keys == nil {
```

Comment:
Might be worth disambiguating "nil because I don't have it" from "nil as in no keys" to be a little more precise here. `nil` is different from an empty slice, even though both have zero length and I'm not completely sure what fetchKeysFromDigest will return in all situations. This might avoid extra db lookups on missing MD5s.

```suggestion
		keys, ok := digests[digest]
		if !ok {
```

Reply:
resolved in [875ccf6](https://github.com/hockeypuck/hockeypuck/pull/186/commits/875ccf656d20df48c87b8ccd44065bec7998f0f2)